File: intro_scala/book-exercises/101-Fraction.scala
//Reduce a fraction to its lowest terms

/*
Compute the greatests common divisor of two integers
@param n the first integer under consideration 
@param m the second integer under consideration
@return the greatest commmon divisor of the integers
*/

import scala.math._
import io.StdIn._

def gcd(n:Int,m:Int):Int ={
    //Inicialize d to the smaller of n and m
    var d= min(m,n)
    while ( n%d!=0 | m%d != 0) {
        d=d-1
    }
    d
}

def reduce(num:Int, den:Int):(Int,Int) = {
    //If the numerator is 0 then the reduced fraction is 0/1
    if (num == 0) (0,1) else {
        val g = gcd(num,den)
        (floor(num/g).toInt,floor(den/g).toInt)
    } 
}

@main 
def main():Unit = {
    //Read the input from user
    println("Enter the numerator: ")
    val num = readInt
    println("Enter the denominator: ")
    val den = readInt

    //Compute the reduced fraction 
    val (n,d) = reduce(num,den)

    println(num+"/"+den+" can be reduced to "+n+"/"+d)
}

File: intro_scala/book-exercises/101-Fraction.py
#Reduce a fraction to its lowest terms 

'''
Compute the greatests common divisor of two integers
@param n the first integer under consideration (must be non-zero)
@param m the second integer under consideration (must be non-zero)
@return the greatest common divisor of the integers
'''


def gcd(n,m):
    #Inicialize d to the smaller of n and m
    d = min(m,n)
    #Use a while loop to find the greatest common divisor 
    # of n and m
    while n % d != 0 or m %d !=0:
        d=d-1
    return d

'''
Reduce a fraction to lowest terms 
@param the integer numerator of the fraction 
@param the integer denominator of the fraction (non-zero)
@return the numerator and denominator of the reduced fraction

'''


def reduce(num, den):
    #if the numerator is 0 then the reduced fracion is 0/1
    if num == 0:
        return (0,1)
    #Compute the greatest common divisor of the numerator and
    #denominator
    g = gcd(num,den)
    #Divide both the numerator and denominator
    #by the gcd and return the result
    return (num//g,den//g)


#Read the fraction from user display and reduce fraction
def main():
    #Read the input from usesr 
    num = int(input("Enter the numerator: "))
    den = int(input("Enter the denominator: "))

    #Compute the reduced fraction 
    (n,d) = reduce(num,den)

    #Display the result 
    print("%d/%d can be reduced to %d/%d." % (num,den,n,d))

#call de main function
main()

